src/main/java/uk/gov/pay/adminusers/model/CreateUserRequest.java

```diff
@@ -42,30 +44,39 @@ public static CreateUserRequest from(String username, String password, String em
-                gatewayAccountIds.sort(numericallyThenLexicographically());
-            }
-            if (node.get(FIELD_SERVICE_EXTERNAL_IDS) != null) {
-                node.get(FIELD_SERVICE_EXTERNAL_IDS)
-                        .iterator()
-                        .forEachRemaining((serviceExternalIdNode) -> serviceExternalIds.add(serviceExternalIdNode.asText()));
-            }
-            String username = node.get(FIELD_USERNAME).asText();
-            String password = getOrElseRandom(node.get(FIELD_PASSWORD), randomUuid());
-            String email = node.get(FIELD_EMAIL).asText();
-            String telephoneNumber = node.get(FIELD_TELEPHONE_NUMBER).asText();
-            String otpKey = getOrElseRandom(node.get(FIELD_OTP_KEY), newId());
-            String features = getOrElseRandom(node.get(FIELD_FEATURES), null);
-            return from(username, password, email, gatewayAccountIds, serviceExternalIds, otpKey, telephoneNumber, features);
-        } catch (NullPointerException e) {
-            throw new RuntimeException("Error retrieving required fields for creating a user", e);
+        final List<String> gatewayAccountIds = safelyGetList(node, FIELD_GATEWAY_ACCOUNT_IDS);
+        gatewayAccountIds.sort(numericallyThenLexicographically());
+
+        final List<String> serviceExternalIds = safelyGetList(node, FIELD_SERVICE_EXTERNAL_IDS);
+
+        String username = safelyGetRequiredNodeAsText(node, FIELD_USERNAME);
+        String password = getOrElseRandom(node.get(FIELD_PASSWORD), randomUuid());
+        String email = safelyGetRequiredNodeAsText(node, FIELD_EMAIL);
+        String telephoneNumber = safelyGetRequiredNodeAsText(node, FIELD_TELEPHONE_NUMBER);
+        String otpKey = getOrElseRandom(node.get(FIELD_OTP_KEY), newId());
+        String features = getOrElseRandom(node.get(FIELD_FEATURES), null);
+        return from(username, password, email, gatewayAccountIds, serviceExternalIds, otpKey, telephoneNumber, features);
+    }
+
+    private static List<String> safelyGetList(JsonNode node, String fieldName) {
+        if (node == null || node.get(fieldName) == null) {
+            return Collections.emptyList();
         }
+
+        var results = new ArrayList<String>();
+        node.get(fieldName)
+                .iterator()
+                .forEachRemaining(nodeValue -> safelyGetNodeAsText(nodeValue).ifPresent(results::add));
+        return results;
+    }
+
+    private static String safelyGetRequiredNodeAsText(JsonNode node, String fieldName) {
+        return safelyGetNodeAsText(node.get(fieldName))
+                .orElseThrow(() -> new RuntimeException(String.format("Error retrieving field {} for creating a user", fieldName)));
+    }
+
+    private static Optional<String> safelyGetNodeAsText(JsonNode node) {
+        return node == null ? Optional.empty() : Optional.ofNullable(node.asText());
```

Comment:
This could be `return Optional.ofNullable(node).map(JsonNode::asText)`, which is possibly small enough it could just be inlined inside `safelyGetList(…)`:

```java
// …
forEachRemaining(Optional::ofNullable).map(JsonNode::asText).ifPresent(results::add));
// …
```

Reply:
Could do, but I was trying to avoid that as I'm not convinced wrapping values we already have in a new optional to avoid a simple null check is a good idea, but then perhaps its a bit more fluid to read and I should get on board with it. I'll try it on and see how it feels :)